### PR TITLE
Do rollout restart after deploy to ensure latest code is running

### PR DESCRIPTION
**Pull request by dpoirier**

### CODE REVIEW

File: tasks/main.yml [hunk 1]
```diff
@@ -172,3 +172,15 @@
       fail_on_error: yes
       strict: yes
   with_items: "{{ k8s_templates }}"
+
+- name: Do a rollout so we get our services updated even if image tags didn't change
```
Comment (copelco) [inline tasks/main.yml]: If a specific tag was used in the deploy, will this task trigger a 2nd restart of all the pods? (Does that matter?)

**Review by tobiasmcnulty — APPROVED** [review]
🚀 Untested, but looks reasonable to me! It'd be nice to have a way to turn this off (`k8s_rollout_restart`?), or even make it opt-in (i.e., default to `no`?). Using the same tag across deploys does not seem like a supported/expected pattern in K8s, so I feel like we're going to run into other issues with it if we make this the default behavior...

**Comment by tobiasmcnulty** [comment]
Ah, I see @copelco beat me to it. That's exactly what I am concerned about.

**Comment by dpoirier** [comment]
@copelco @tobiasmcnulty Updated so doing a `rollout` is not the default, but a project can set k8s_rollout_after_deploy if they want it.

File: defaults/main.yml [hunk 2]
```diff
@@ -27,6 +27,14 @@ k8s_auth_ssl_ca_cert: ""  # should be a file path (relative to deploy.yaml)
 #       auth: "{{ [docker_hub_user, docker_hub_password] | join(':') | b64encode }}"
 k8s_dockerconfigjson: {}
 
+# Whether to do a "kubectl rollout restart" on each deploy.
+# This should be unnecessary if deploys are using unique image tags,
+# but might be desired if a project is using tags like 'latest'
+# or 'staging' that will refer to different tags at different times,
+# to make sure the latest version of the image with that tag gets
+# pulled and used.
```
Comment (tobiasmcnulty) [inline defaults/main.yml]:
```suggestion
# pulled and used. When setting this 'true', ensure you also set
# `k8s_container_image_pull_policy: Always`
```

File: defaults/main.yml [hunk 3]
```diff
@@ -27,6 +27,14 @@ k8s_auth_ssl_ca_cert: ""  # should be a file path (relative to deploy.yaml)
...
+k8s_rollout_after_deploy: False
```
Comment (tobiasmcnulty) [inline defaults/main.yml:37]: For consistency with the rest of this file:
```suggestion
k8s_rollout_after_deploy: false
```

File: CHANGES.rst [hunk 4]
```diff
@@ -8,10 +8,10 @@ Changes
 Next release
 ------------
 
-* Use rollout to ensure that pods are restarted when we deploy. This ensures
-  that even if our image tag is unchanged (like if we're using a branch name),
-  we'll still pull the latest image with that tag and be running it when the
-  deploy completes.
+* If k8s_rollout_after_deploy is True, use rollout to ensure that pods are restarted
```
Comment (tobiasmcnulty) [inline CHANGES.rst]:
```suggestion
* If ``k8s_rollout_after_deploy`` is ``true``, use rollout to ensure that pods are restarted
```

File: tasks/main.yml [hunk 5]
```diff
@@ -173,14 +173,16 @@
       strict: yes
   with_items: "{{ k8s_templates }}"
 
-- name: Do a rollout so we get our services updated even if image tags didn't change
-  command: kubectl --namespace "{{ k8s_namespace }}" rollout restart deployment/{{ item["name"] }}
-  with_items: "{{ k8s_web_containers }}"
+- when: k8s_rollout_after_deploy
```
Comment (tobiasmcnulty) [inline tasks/main.yml:176]: ❤️

**Review by tobiasmcnulty — APPROVED** [review]
🚢

**Pull request merged**